code/datums/cleaning.dm

```diff
@@ -0,0 +1,81 @@
+//For handling standard click-to-clean items like soap and mops.
+/datum/proc/cleaning_act(atom/target, mob/user, atom/cleaner, cleanspeed = 50, ismop = FALSE)
+	var/innatecleaner = FALSE //If the user cleaning is innately able to clean, i.e. Lusty Xenomorph Maid
+	var/cmag_cleantime = 50 //The cleaning time for cmagged objects is locked to this, for balance reasons
+
+	if(user == cleaner)
+		innatecleaner = TRUE
+
+	if(ismop)
+		if(cleaner.reagents.total_volume < 1)
+			to_chat(user, "<span class='warning'>Your mop is dry!</span>")
+			return
+
+	if(HAS_TRAIT(target, TRAIT_CMAGGED))
+		user.visible_message("<span class='notice'>[user] starts to clean the ooze off \the [target.name].</span>", "<span class='notice'>You start to clean the ooze off \the [target.name].</span>")
+		if(do_after(user, cmag_cleantime, target = target) && target)
+			user.visible_message("<span class='notice'>[user] cleans the ooze off \the [target.name].</span>", "<span class='notice'>You clean the ooze off \the [target.name].</span>")
+			REMOVE_TRAIT(target, TRAIT_CMAGGED, "clown_emag")
+		return
+
+	if(user.client && (target in user.client.screen)) //You can't clean items you're wearing for technical reasons
+		to_chat(user, "<span class='notice'>You need to take that [target.name] off before cleaning it.</span>")
+		return
+
+	if(istype(target, /obj/effect/decal/cleanable) || istype(target, /obj/effect/rune))
+		if(!innatecleaner)
+			user.visible_message("<span class='warning'>[user] begins to scrub \the [target.name] out with [src].</span>")
+		else
+			user.visible_message("<span class='warning'>[user] begins to scrub out \the [target.name].</span>")
+		if(do_after(user, cleanspeed, target = target) && target)
```

Comment:
you do this very `do_after` multiple times in this proc, couldn't you do if(!do_after(...)) => return somewhere near the top of the proc as a general thing